File: web/src/lib/crypto/golden-vectors.test.ts
// @vitest-environment node
// The client-crypto contract: every assertion pins this implementation
// to the repo-root client-crypto-vectors.json exported from the Rust
// reference (`make vectors`). Mirrors android/core/crypto
// GoldenVectorsTest.kt, plus the two fields the Kotlin test leaves
// unused (keyBackup.contentKeyHex / plaintextHex).

import { readFileSync } from "node:fs";
import { Buffer } from "node:buffer";
import { expect, it } from "vitest";
import { fromHex, toHex } from "./bytes";
import { CborEncoder } from "./cbor";
import { commitment, preDigest, sha256Tagged, Tags, verify } from "./hashing";
import { ActId, NodeId, parseFamily } from "./identifiers";
import { canonicalDeps, preCommitmentMsg, Proposal, StructuralBody, VerifiedAct } from "./handshake";
import { ActorKey } from "./actor-key";
import { decodeProposal, decodeVerifiedAct, encodePreCommitmentOf, encodeProposal, encodeVerifiedAct } from "./wire";
import { openKeyBackup, RecoveryCode, sealKeyBackup, signUpload, UPLOAD_PROOF_TAG } from "./key-backup";

interface BodyVector {
  actId: string;
  assertedParents: string[];
  author: string;
  canonicalBytesHex: string;
  family: string;
  license: string | null;
  middle: string | null;
  pD: number;
  pI: number;
  seq: number;
  settlementRef: string | null;
  target: string;
}

interface Vectors {
  version: number;
  encoding: { cborHex: string; value: string }[];
  sha256Tagged: { digestHex: string; partsHex: string[]; tagUtf8: string }[];
  signing: {
    l0Address: string;
    publicKeyHex: string;
    seedHex: string;
    samples: { msgUtf8: string; signatureHex: string; tagUtf8: string }[];
  };
  structuralBodies: BodyVector[];
  handshake: {
    approvalActId: string;
    approvalSignatureHex: string;
    canonicalDepsHex: string;
    contentCommitmentHex: string;
    contentPreDigestHex: string;
    contentSaltHex: string;
    depsCommitmentHex: string;
    depsPreDigestHex: string;
    depsSaltHex: string;
    host: { publicKeyHex: string; seedHex: string };
    hostSealHex: string;
    nonceHex: string;
    preCommitmentMsgHex: string;
    preSignatureHex: string;
    proposal: { body: BodyVector; deps: string[]; payloadHex: string };
    sealMsgHex: string;
    wirePreCommitmentHex: string;
    wireProposalHex: string;
    wireVerifiedActHex: string;
  };
  keyBackup: {
    aesNonceHex: string;
    blobBase64: string;
    blobHex: string;
    contentKeyHex: string;
    hkdfInfoUtf8: string;
    hkdfSaltHex: string;
    plaintextHex: string;
    recoveryCodeBytesHex: string;
    recoveryCodeDisplay: string;
    uploadChallengeHex: string;
    uploadProofTagUtf8: string;
    uploadSignatureHex: string;
  };
}

const vectors = JSON.parse(
  readFileSync(new URL("../../../../client-crypto-vectors.json", import.meta.url), "utf-8"),
) as Vectors;

const utf8 = (s: string) => Uint8Array.from(new TextEncoder().encode(s));

function bodyOf(v: BodyVector): StructuralBody {
  return new StructuralBody({
    author: v.author,
    seq: BigInt(v.seq),
    family: parseFamily(v.family),
    middle: v.middle === null ? null : NodeId.parse(v.middle),
    target: NodeId.parse(v.target),
    pD: v.pD,
    pI: v.pI,
    settlementRef: v.settlementRef === null ? null : ActId.parse(v.settlementRef),
    license: v.license,
    assertedParents: v.assertedParents.map((p) => ActId.parse(p)),
  });
}

it("the vectors file is current", () => {
  expect(vectors.version).toBe(1);
});

it("encoding vectors match", () => {
  const produced: { value: string; bytes: Uint8Array }[] = [
    { value: "uint 0", bytes: new CborEncoder().uint(0n).finish() },
    { value: "uint 23", bytes: new CborEncoder().uint(23n).finish() },
    { value: "uint 24", bytes: new CborEncoder().uint(24n).finish() },
    { value: "uint 255", bytes: new CborEncoder().uint(255n).finish() },
    { value: "uint 256", bytes: new CborEncoder().uint(256n).finish() },
    { value: "uint 65535", bytes: new CborEncoder().uint(65535n).finish() },
    { value: "uint 65536", bytes: new CborEncoder().uint(65536n).finish() },
    { value: "uint 4294967295", bytes: new CborEncoder().uint(4294967295n).finish() },
    { value: "uint 4294967296", bytes: new CborEncoder().uint(4294967296n).finish() },
    { value: "bytes ''", bytes: new CborEncoder().bytes(new Uint8Array(0)).finish() },
    { value: "bytes 0102", bytes: new CborEncoder().bytes(Uint8Array.of(1, 2)).finish() },
    { value: "text ''", bytes: new CborEncoder().text("").finish() },
    { value: "text 'cogra'", bytes: new CborEncoder().text("cogra").finish() },
    { value: "text 'héllo → ✓'", bytes: new CborEncoder().text("héllo → ✓").finish() },
    { value: "float 0.0", bytes: new CborEncoder().float(0.0).finish() },
    { value: "float 1.0", bytes: new CborEncoder().float(1.0).finish() },
    { value: "float -1.0", bytes: new CborEncoder().float(-1.0).finish() },
    { value: "float 0.5", bytes: new CborEncoder().float(0.5).finish() },
    { value: "float -0.25", bytes: new CborEncoder().float(-0.25).finish() },
    { value: "null", bytes: new CborEncoder().nul().finish() },
    { value: "array(0)", bytes: new CborEncoder().array(0n).finish() },
    {
      value: "array(2)[uint 1, array(1)[text 'x']]",
      bytes: new CborEncoder().array(2n).uint(1n).array(1n).text("x").finish(),
    },
  ];
  expect(produced.length).toBe(vectors.encoding.length);
  for (let i = 0; i < produced.length; i++) {
    expect(produced[i].value).toBe(vectors.encoding[i].value);
    expect(toHex(produced[i].bytes)).toBe(vectors.encoding[i].cborHex);
  }
});

it("tagged hash vectors match", async () => {
  for (const v of vectors.sha256Tagged) {
    const digest = await sha256Tagged(v.tagUtf8, v.partsHex.map(fromHex));
    expect(toHex(digest)).toBe(v.digestHex);
  }
});

it("signing vectors match", async () => {
  const s = vectors.signing;
  const key = await ActorKey.fromSeed(fromHex(s.seedHex));
  expect(toHex(key.publicKeyBytes())).toBe(s.publicKeyHex);
  expect(await key.address()).toBe(s.l0Address);
  for (const sample of s.samples) {
    // Ed25519 is deterministic: byte-equality, not just verification.
    const signature = await key.signTagged(sample.tagUtf8, utf8(sample.msgUtf8));
    expect(toHex(signature)).toBe(sample.signatureHex);
    expect(await verify(key.publicKeyBytes(), sample.tagUtf8, utf8(sample.msgUtf8), signature)).toBe(true);
  }
});

it("structural body vectors match", () => {
  for (const v of vectors.structuralBodies) {
    const body = bodyOf(v);
    expect(body.actId().toString()).toBe(v.actId);
    expect(toHex(body.canonicalBytes())).toBe(v.canonicalBytesHex);
  }
});

it("handshake vectors match", async () => {
  const h = vectors.handshake;
  const actor = await ActorKey.fromSeed(fromHex(vectors.signing.seedHex));
  const proposal = new Proposal(
    bodyOf(h.proposal.body),
    fromHex(h.proposal.payloadHex),
    h.proposal.deps.map((d) => ActId.parse(d)),
  );
  expect(toHex(canonicalDeps(proposal.deps))).toBe(h.canonicalDepsHex);
  expect(toHex(encodeProposal(proposal))).toBe(h.wireProposalHex);
  expect(decodeProposal(fromHex(h.wireProposalHex)).equals(proposal)).toBe(true);

  const nonce = fromHex(h.nonceHex);
  const pre = await actor.preSign(proposal, nonce);
  expect(toHex(await preDigest(Tags.PRE_DIGEST_CONTENT, nonce, proposal.payload))).toBe(h.contentPreDigestHex);
  expect(toHex(await preDigest(Tags.PRE_DIGEST_DEPS, nonce, canonicalDeps(proposal.deps)))).toBe(
    h.depsPreDigestHex,
  );
  expect(
    toHex(preCommitmentMsg(proposal.body, fromHex(h.contentPreDigestHex), fromHex(h.depsPreDigestHex))),
  ).toBe(h.preCommitmentMsgHex);
  expect(toHex(pre.preSignature)).toBe(h.preSignatureHex);
  expect(toHex(encodePreCommitmentOf(pre))).toBe(h.wirePreCommitmentHex);

  const contentSalt = fromHex(h.contentSaltHex);
  const depsSalt = fromHex(h.depsSaltHex);
  const sealed = new VerifiedAct({
    proposal,
    authorPubkey: actor.publicKeyBytes(),
    nonce,
    preSignature: pre.preSignature,
    contentSalt,
    depsSalt,
    contentCommitment: fromHex(h.contentCommitmentHex),
    depsCommitment: fromHex(h.depsCommitmentHex),
    hostSeal: fromHex(h.hostSealHex),
  });
  expect(toHex(await commitment(Tags.COMMIT_CONTENT, contentSalt, proposal.payload))).toBe(
    h.contentCommitmentHex,
  );
  expect(toHex(await commitment(Tags.COMMIT_DEPS, depsSalt, canonicalDeps(proposal.deps)))).toBe(
    h.depsCommitmentHex,
  );
  expect(toHex(sealed.sealMsg())).toBe(h.sealMsgHex);
  expect(toHex(encodeVerifiedAct(sealed))).toBe(h.wireVerifiedActHex);
  const decoded = decodeVerifiedAct(fromHex(h.wireVerifiedActHex));
  expect(decoded.proposal.equals(proposal)).toBe(true);
  expect(toHex(decoded.hostSeal)).toBe(h.hostSealHex);

  const witness = await actor.approve(pre, sealed, fromHex(h.host.publicKeyHex));
  expect(witness.actId.toString()).toBe(h.approvalActId);
  expect(toHex(witness.approvalSignature)).toBe(h.approvalSignatureHex);
});

it("key backup vectors match", async () => {
  const k = vectors.keyBackup;
  const seed = fromHex(vectors.signing.seedHex);
  const code = new RecoveryCode(fromHex(k.recoveryCodeBytesHex));
  expect(code.display()).toBe(k.recoveryCodeDisplay);
  expect(k.hkdfInfoUtf8).toBe("cogra:key-backup:v1");

  const blob = await sealKeyBackup(seed, code, fromHex(k.hkdfSaltHex), fromHex(k.aesNonceHex));
  expect(toHex(blob)).toBe(k.blobHex);
  expect(Buffer.from(blob).toString("base64")).toBe(k.blobBase64);

  const retyped = RecoveryCode.fromInput(k.recoveryCodeDisplay);
  expect(toHex(retyped.bytes)).toBe(toHex(code.bytes));
  expect(toHex(await openKeyBackup(fromHex(k.blobHex), retyped))).toBe(vectors.signing.seedHex);

  // The two fields the Kotlin golden test leaves unused.
  expect(toHex(new CborEncoder().array(2n).bytes(seed).uint(1n).finish())).toBe(k.plaintextHex);
  const ikm = await crypto.subtle.importKey("raw", fromHex(k.recoveryCodeBytesHex), "HKDF", false, [
    "deriveBits",
  ]);
  const contentKey = await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt: fromHex(k.hkdfSaltHex), info: utf8(k.hkdfInfoUtf8) },
    ikm,
    256,
  );
  expect(toHex(new Uint8Array(contentKey))).toBe(k.contentKeyHex);
});

it("key-backup upload proof vectors match", async () => {
  const k = vectors.keyBackup;
  expect(UPLOAD_PROOF_TAG).toBe(k.uploadProofTagUtf8);

  const actor = await ActorKey.fromSeed(fromHex(vectors.signing.seedHex));
  const signature = await signUpload(
    actor,
    fromHex(k.uploadChallengeHex),
    fromHex(k.blobHex),
  );
  expect(toHex(signature)).toBe(k.uploadSignatureHex);
});

File: android/core/crypto/src/test/kotlin/com/cogra/crypto/GoldenVectorsTest.kt
// Pins this module to `client-crypto-vectors.json` (repo root) — the
// cross-language golden vectors exported from the Rust reference
// (`make vectors`). Every group is exercised: the deterministic CBOR
// subset, the tagged hashing and Ed25519 signing, the structural bodies,
// the complete admission handshake, and the key-backup blob.

package com.cogra.crypto

import com.google.common.truth.Truth.assertThat
import java.io.File
import java.util.Base64
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Test

private val vectors: JsonElement by lazy {
    // The repo-root contract file, exactly as the Rust exporter wrote it.
    Json.parseToJsonElement(File("../../../client-crypto-vectors.json").readText())
}

private fun JsonElement.str(key: String): String =
    jsonObject.getValue(key).jsonPrimitive.content

private fun JsonElement.strOrNull(key: String): String? =
    jsonObject.getValue(key).let { if (it is JsonNull) null else it.jsonPrimitive.content }

private fun bodyOf(json: JsonElement): StructuralBody = StructuralBody(
    author = json.str("author"),
    seq = json.str("seq").toULong(),
    family = Family.parse(json.str("family")),
    middle = json.strOrNull("middle")?.let { NodeId.parse(it) },
    target = NodeId.parse(json.str("target")),
    pD = json.str("pD").toDouble(),
    pI = json.str("pI").toDouble(),
    settlementRef = json.strOrNull("settlementRef")?.let { ActId.parse(it) },
    license = json.strOrNull("license"),
    assertedParents = json.jsonObject.getValue("assertedParents").jsonArray
        .map { ActId.parse(it.jsonPrimitive.content) },
)

class GoldenVectorsTest {

    @Test
    fun vectorsFileIsCurrent() {
        assertThat(vectors.str("version")).isEqualTo("1")
    }

    @Test
    fun encodingVectorsMatch() {
        // The same 22 values the exporter encodes, in the same order.
        val produced: List<Pair<String, ByteArray>> = listOf(
            "uint 0" to CborEncoder().uint(0u).finish(),
            "uint 23" to CborEncoder().uint(23u).finish(),
            "uint 24" to CborEncoder().uint(24u).finish(),
            "uint 255" to CborEncoder().uint(255u).finish(),
            "uint 256" to CborEncoder().uint(256u).finish(),
            "uint 65535" to CborEncoder().uint(65535u).finish(),
            "uint 65536" to CborEncoder().uint(65536u).finish(),
            "uint 4294967295" to CborEncoder().uint(4_294_967_295u).finish(),
            "uint 4294967296" to CborEncoder().uint(4_294_967_296u).finish(),
            "bytes ''" to CborEncoder().bytes(byteArrayOf()).finish(),
            "bytes 0102" to CborEncoder().bytes(byteArrayOf(1, 2)).finish(),
            "text ''" to CborEncoder().text("").finish(),
            "text 'cogra'" to CborEncoder().text("cogra").finish(),
            "text 'héllo → ✓'" to CborEncoder().text("héllo → ✓").finish(),
            "float 0.0" to CborEncoder().float(0.0).finish(),
            "float 1.0" to CborEncoder().float(1.0).finish(),
            "float -1.0" to CborEncoder().float(-1.0).finish(),
            "float 0.5" to CborEncoder().float(0.5).finish(),
            "float -0.25" to CborEncoder().float(-0.25).finish(),
            "null" to CborEncoder().nul().finish(),
            "array(0)" to CborEncoder().array(0u).finish(),
            "array(2)[uint 1, array(1)[text 'x']]" to
                CborEncoder().array(2u).uint(1u).array(1u).text("x").finish(),
        )
        val expected = vectors.jsonObject.getValue("encoding").jsonArray
        assertThat(produced.size).isEqualTo(expected.size)
        for ((case, vector) in produced.zip(expected)) {
            assertThat(vector.str("value")).isEqualTo(case.first)
            assertThat(case.second.toHex()).isEqualTo(vector.str("cborHex"))
        }
    }

    @Test
    fun taggedHashVectorsMatch() {
        for (case in vectors.jsonObject.getValue("sha256Tagged").jsonArray) {
            val parts = case.jsonObject.getValue("partsHex").jsonArray
                .map { it.jsonPrimitive.content.hexToBytes() }
            assertThat(sha256Tagged(case.str("tagUtf8"), parts).toHex())
                .isEqualTo(case.str("digestHex"))
        }
    }

    @Test
    fun signingVectorsMatch() {
        val signing = vectors.jsonObject.getValue("signing")
        val key = ActorKey.fromSeed(signing.str("seedHex").hexToBytes())
        assertThat(key.publicKeyBytes().toHex()).isEqualTo(signing.str("publicKeyHex"))
        assertThat(key.address()).isEqualTo(signing.str("l0Address"))
        for (sample in signing.jsonObject.getValue("samples").jsonArray) {
            val tag = sample.str("tagUtf8")
            val msg = sample.str("msgUtf8").toByteArray(Charsets.UTF_8)
            // Ed25519 is deterministic: byte-equality, not just verification.
            val signature = key.signTagged(tag, msg)
            assertThat(signature.toHex()).isEqualTo(sample.str("signatureHex"))
            assertThat(verify(key.publicKeyBytes(), tag, msg, signature)).isTrue()
        }
    }

    @Test
    fun structuralBodyVectorsMatch() {
        for (case in vectors.jsonObject.getValue("structuralBodies").jsonArray) {
            val body = bodyOf(case)
            assertThat(body.actId().toString()).isEqualTo(case.str("actId"))
            assertThat(body.canonicalBytes().toHex()).isEqualTo(case.str("canonicalBytesHex"))
        }
    }

    @Test
    fun handshakeVectorsMatch() {
        val hs = vectors.jsonObject.getValue("handshake")
        val actor = ActorKey.fromSeed(vectors.jsonObject.getValue("signing").str("seedHex").hexToBytes())
        val hostPubkey = hs.jsonObject.getValue("host").str("publicKeyHex").hexToBytes()

        val proposalJson = hs.jsonObject.getValue("proposal")
        val proposal = Proposal(
            body = bodyOf(proposalJson.jsonObject.getValue("body")),
            payload = proposalJson.str("payloadHex").hexToBytes(),
            deps = proposalJson.jsonObject.getValue("deps").jsonArray
                .map { ActId.parse(it.jsonPrimitive.content) },
        )
        assertThat(canonicalDeps(proposal.deps).toHex()).isEqualTo(hs.str("canonicalDepsHex"))
        assertThat(encodeProposal(proposal).toHex()).isEqualTo(hs.str("wireProposalHex"))
        assertThat(decodeProposal(hs.str("wireProposalHex").hexToBytes())).isEqualTo(proposal)

        // Pre-sign under the pinned nonce: every intermediate matches.
        val nonce = hs.str("nonceHex").hexToBytes()
        val pre = actor.preSignWith(proposal, nonce)
        assertThat(preDigest(Tags.PRE_DIGEST_CONTENT, nonce, proposal.payload).toHex())
            .isEqualTo(hs.str("contentPreDigestHex"))
        assertThat(preDigest(Tags.PRE_DIGEST_DEPS, nonce, canonicalDeps(proposal.deps)).toHex())
            .isEqualTo(hs.str("depsPreDigestHex"))
        assertThat(
            preCommitmentMsg(
                proposal.body,
                hs.str("contentPreDigestHex").hexToBytes(),
                hs.str("depsPreDigestHex").hexToBytes(),
            ).toHex()
        ).isEqualTo(hs.str("preCommitmentMsgHex"))
        assertThat(pre.preSignature.toHex()).isEqualTo(hs.str("preSignatureHex"))
        assertThat(encodePreCommitmentOf(pre).toHex()).isEqualTo(hs.str("wirePreCommitmentHex"))

        // The sealed act: commitments recompute, the seal message and
        // wire form match, and the host seal verifies.
        val sealed = VerifiedAct(
            proposal = proposal,
            authorPubkey = actor.publicKeyBytes(),
            nonce = nonce,
            preSignature = pre.preSignature,
            contentSalt = hs.str("contentSaltHex").hexToBytes(),
            depsSalt = hs.str("depsSaltHex").hexToBytes(),
            contentCommitment = hs.str("contentCommitmentHex").hexToBytes(),
            depsCommitment = hs.str("depsCommitmentHex").hexToBytes(),
            hostSeal = hs.str("hostSealHex").hexToBytes(),
        )
        assertThat(
            commitment(Tags.COMMIT_CONTENT, sealed.contentSalt, proposal.payload).toHex()
        ).isEqualTo(hs.str("contentCommitmentHex"))
        assertThat(
            commitment(Tags.COMMIT_DEPS, sealed.depsSalt, canonicalDeps(proposal.deps)).toHex()
        ).isEqualTo(hs.str("depsCommitmentHex"))
        assertThat(sealed.sealMsg().toHex()).isEqualTo(hs.str("sealMsgHex"))
        assertThat(encodeVerifiedAct(sealed).toHex()).isEqualTo(hs.str("wireVerifiedActHex"))
        // Every field, not a sample of two: a decoder that swapped the
        // two same-length salts passed the old pair of assertions.
        assertThat(decodeVerifiedAct(hs.str("wireVerifiedActHex").hexToBytes())).isEqualTo(sealed)

        // The real approval path accepts the pinned act and reproduces
        // the pinned witness.
        val witness = actor.approve(pre, sealed, hostPubkey)
        assertThat(witness.actId.toString()).isEqualTo(hs.str("approvalActId"))
        assertThat(witness.approvalSignature.toHex()).isEqualTo(hs.str("approvalSignatureHex"))
    }

    @Test
    fun keyBackupVectorsMatch() {
        val kb = vectors.jsonObject.getValue("keyBackup")
        val seed = vectors.jsonObject.getValue("signing").str("seedHex").hexToBytes()
        val code = RecoveryCode(kb.str("recoveryCodeBytesHex").hexToBytes())

        assertThat(code.display()).isEqualTo(kb.str("recoveryCodeDisplay"))
        assertThat(kb.str("hkdfInfoUtf8")).isEqualTo("cogra:key-backup:v1")

        val blob = sealKeyBackupWith(
            seed = seed,
            code = code,
            salt = kb.str("hkdfSaltHex").hexToBytes(),
            nonce = kb.str("aesNonceHex").hexToBytes(),
        )
        assertThat(blob.toHex()).isEqualTo(kb.str("blobHex"))
        assertThat(Base64.getEncoder().encodeToString(blob)).isEqualTo(kb.str("blobBase64"))

        // The blob opens under the display form a user would retype.
        val retyped = RecoveryCode.fromInput(kb.str("recoveryCodeDisplay"))
        assertThat(retyped.bytes()).isEqualTo(code.bytes())
        assertThat(openKeyBackup(kb.str("blobHex").hexToBytes(), retyped)).isEqualTo(seed)
    }

    @Test
    fun keyBackupUploadProofVectorsMatch() {
        val kb = vectors.jsonObject.getValue("keyBackup")
        assertThat(UPLOAD_PROOF_TAG).isEqualTo(kb.str("uploadProofTagUtf8"))

        val actor = ActorKey.fromSeed(vectors.jsonObject.getValue("signing").str("seedHex").hexToBytes())
        val signature = signUpload(
            key = actor,
            challenge = kb.str("uploadChallengeHex").hexToBytes(),
            blob = kb.str("blobHex").hexToBytes(),
        )
        assertThat(signature.toHex()).isEqualTo(kb.str("uploadSignatureHex"))
    }
}
